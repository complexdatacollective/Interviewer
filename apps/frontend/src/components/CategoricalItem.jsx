/* eslint-disable jsx-a11y/click-events-have-key-events */
import React from 'react';
import { compose, withProps, withState } from 'recompose';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { MarkdownLabel } from '@codaco/ui';
import { DropTarget, MonitorDropTarget } from '../behaviours/DragAndDrop';
import NodeList from './NodeList';

const Flipped = ({ children }) => (<div>{children}</div>);

/**
  * Renders a droppable CategoricalBin item
  */
const CategoricalItem = ({
  accentColor,
  details,
  id,
  isExpanded,
  isOver,
  label,
  nodes,
  onClick,
  onClickItem,
  sortOrder,
  willAccept,
  stage,
}) => {
  const classNames = cx(
    'categorical-item',
    { 'categorical-item--hover': willAccept && isOver },
    { 'categorical-item--expanded': isExpanded },
  );

  return (
    <Flipped flipId={id}>
      <div
        className={classNames}
        style={{ '--categorical-item-color': accentColor }}
        onClick={onClick}
      >
        <div className="categorical-item__disk" />
        <div className="categorical-item__inner">
          <Flipped inverseFlipId={id} scale>
            <div className="categorical-item__title">
              <h3>
                <MarkdownLabel inline label={label} />
              </h3>
              {(!isExpanded && details)
                && (
                  <h5>
                    <MarkdownLabel inline label={details} />
                  </h5>
                )}
            </div>
          </Flipped>
          {isExpanded
            && (
              <div className="categorical-item__content">
                <NodeList
                  stage={stage}
                  listId={`CATBIN_NODE_LIST_${label}`}
                  id={`CATBIN_NODE_LIST_${label}`}
                  onItemClick={onClickItem}
                  items={nodes}
                  sortOrder={sortOrder}
                />
              </div>
            )}
        </div>
      </div>
    </Flipped>
  );
};

CategoricalItem.propTypes = {
  accentColor: PropTypes.string,
  details: PropTypes.string,
  id: PropTypes.string.isRequired,
  isExpanded: PropTypes.bool,
  isOver: PropTypes.bool,
  label: PropTypes.string,
  nodes: PropTypes.array,
  onClick: PropTypes.func,
  onClickItem: PropTypes.func,
  sortOrder: PropTypes.array,
  willAccept: PropTypes.bool,
};

CategoricalItem.defaultProps = {
  accentColor: 'black',
  details: '',
  isExpanded: false,
  isOver: false,
  label: 'undefined',
  nodes: [],
  onClick: () => { },
  onClickItem: () => { },
  sortOrder: [],
  willAccept: false,
};

export { CategoricalItem as UnconnectedCategoricalItem };

export default compose(
  withState('recentNode', 'setRecentNode', {}),
  withProps((props) => ({
    accepts: () => true,
    onDrop: ({ meta }) => {
      props.onDrop({ meta });
      props.setRecentNode(meta);
    },
  })),
  DropTarget,
  MonitorDropTarget(['isOver', 'willAccept']),
)(CategoricalItem);
