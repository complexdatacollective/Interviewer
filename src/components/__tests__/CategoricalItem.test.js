/* eslint-env jest */


import React from 'react';
import { shallow } from 'enzyme';
import { UnconnectedCategoricalItem as CategoricalItem } from '../CategoricalItem';

describe('CategoricalItem component', () => {
  it('renders unexpanded categorical item', () => {
    const component = shallow(
      <CategoricalItem />,
    );

    expect(component).toMatchSnapshot();

    expect(component.find('.categorical-item').at(0).hasClass('categorical-item--expanded')).toBe(false);
    expect(component.find('.categorical-item__content')).toHaveLength(0);
  });

  it('renders expanded categorical item', () => {
    const component = shallow(
      <CategoricalItem isExpanded />,
    );

    expect(component.find('.categorical-item').at(0).hasClass('categorical-item--expanded')).toBe(true);
    expect(component.find('.categorical-item__content')).toHaveLength(1);
  });

  it('registers click', () => {
    const clickDummy = jest.fn();

    const component = shallow(
      <CategoricalItem onClick={clickDummy} />,
    );

    expect(clickDummy.mock.calls.length).toBe(0);
    component.find('.categorical-item').simulate('click');
    expect(clickDummy.mock.calls.length).toBe(1);
  });
});
