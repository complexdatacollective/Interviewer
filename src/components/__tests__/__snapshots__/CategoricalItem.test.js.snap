// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoricalItem component renders unexpanded categorical item 1`] = `
<FlippedWithContext>
  <div
    className="categorical-item"
    onClick={[Function]}
    style={
      Object {
        "--categorical-item-color": "black",
      }
    }
  >
    <div
      className="categorical-item__disk"
    />
    <div
      className="categorical-item__inner"
    >
      <FlippedWithContext
        scale={true}
      >
        <div
          className="categorical-item__title"
        >
          <h3>
            <Memo(MarkdownLabel)
              inline={true}
              label="undefined"
            />
          </h3>
        </div>
      </FlippedWithContext>
    </div>
  </div>
</FlippedWithContext>
`;
